simple rust package fails to build

Building a rust package with a dependency fails:

```
error: builder for '/nix/store/nzb7yln59r767bx7avrdpyky85g724p2-dream2nix-bug-0.1.0.drv' failed with exit code 101;
       last 10 log lines:
       > Executing cargoSetupPostUnpackHook
       > Finished cargoSetupPostUnpackHook
       > patching sources
       > configuring
       > building
       > Executing cargoBuildHook
       > ++ env CC_x86_64-unknown-linux-gnu=/nix/store/pja9g36cy32z3d51942jqk91a6l2d5nv-gcc-wrapper-10.3.0/bin/cc CXX_x86_64-unknown-linux-gnu=/nix/store/pja9g36cy32z3d51942jqk91a6l2d5nv-gcc-wrapper-10.3.0/bin/c++ CC_x86_64-unknown-linux-gnu=/nix/store/pja9g36cy32z3d51942jqk91a6l2d5nv-gcc-wrapper-10.3.0/bin/cc CXX_x86_64-unknown-linux-gnu=/nix/store/pja9g36cy32z3d51942jqk91a6l2d5nv-gcc-wrapper-10.3.0/bin/c++ cargo build -j 12 --target x86_64-unknown-linux-gnu --frozen --release --package dream2nix-bug
       > error: no matching package named `chrono` found
       > location searched: registry `crates-io`
       > required by package `dream2nix-bug v0.1.0 (/build/4n4yrgfjhhw8ldcqrv8w08gdnr948w0q-source)`
       For full logs, run 'nix log /nix/store/nzb7yln59r767bx7avrdpyky85g724p2-dream2nix-bug-0.1.0.drv'.
```

Repo for repro: https://github.com/lelongg/dream2nix-bug

#130 fixed the initial issue but a new one is reproducible at [the same repo](https://github.com/lelongg/dream2nix-bug).

The error seems to come from `actix_tls` not having the `connect` feature activated.

```
error[E0432]: unresolved imports `actix_tls::connect::new_connector`, `actix_tls::connect::Connect`
  --> /build/4y8ml4p65982cqjdjjzp3by17lsmgrqn-source/../nix-vendor/actix-http-3.0.0-beta.5/src/client/connector.rs:17:5
   |
17 |     new_connector, Connect as TcpConnect, ConnectError as TcpConnectError,
   |     ^^^^^^^^^^^^^  ^^^^^^^^^^^^^^^^^^^^^ no `Connect` in `connect`
   |     |
   |     no `new_connector` in `connect`
   |
help: a similar name exists in the module
   |
17 |     connector, Connect as TcpConnect, ConnectError as TcpConnectError,
   |     ~~~~~~~~~
help: a similar name exists in the module
   |
17 |     new_connector, Connector as TcpConnect, ConnectError as TcpConnectError,
   |                    ~~~~~~~~~

error[E0432]: unresolved import `actix_tls::connect::Connect`
  --> /build/4y8ml4p65982cqjdjjzp3by17lsmgrqn-source/../nix-vendor/actix-http-3.0.0-beta.5/src/client/mod.rs:14:5
   |
14 |     Connect as TcpConnect, ConnectError as TcpConnectError, Connection as TcpConnection,
   |     -------^^^^^^^^^^^^^^
   |     |
   |     no `Connect` in `connect`
   |     help: a similar name exists in the module: `Connector`

error[E0432]: unresolved import `actix_service::pipeline_factory`
  --> /build/4y8ml4p65982cqjdjjzp3by17lsmgrqn-source/../nix-vendor/actix-http-3.0.0-beta.5/src/service.rs:13:21
   |
13 | use actix_service::{pipeline_factory, IntoServiceFactory, Service, ServiceFactory};
   |                     ^^^^^^^^^^^^^^^^ no `pipeline_factory` in the root

error[E0432]: unresolved import `actix_service::pipeline_factory`
 --> /build/4y8ml4p65982cqjdjjzp3by17lsmgrqn-source/../nix-vendor/actix-http-3.0.0-beta.5/src/h1/service.rs:8:21
  |
8 | use actix_service::{pipeline_factory, IntoServiceFactory, Service, ServiceFactory};
  |                     ^^^^^^^^^^^^^^^^ no `pipeline_factory` in the root

error[E0432]: unresolved import `actix_service::pipeline_factory`
  --> /build/4y8ml4p65982cqjdjjzp3by17lsmgrqn-source/../nix-vendor/actix-http-3.0.0-beta.5/src/h2/service.rs:10:29
   |
10 |     fn_factory, fn_service, pipeline_factory, IntoServiceFactory, Service,
   |                             ^^^^^^^^^^^^^^^^ no `pipeline_factory` in the root

error[E0282]: type annotations needed
   --> /build/4y8ml4p65982cqjdjjzp3by17lsmgrqn-source/../nix-vendor/actix-http-3.0.0-beta.5/src/client/connector.rs:267:28
    |
267 |     pub fn finish(self) -> ConnectorService<S, Io> {
    |                            ^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type for type parameter `Io`

error[E0282]: type annotations needed
   --> /build/4y8ml4p65982cqjdjjzp3by17lsmgrqn-source/../nix-vendor/actix-http-3.0.0-beta.5/src/client/connector.rs:533:13
    |
533 | impl<S, Io> Service<Connect> for TcpConnectorInnerService<S>
    |             ^^^^^^^^^^^^^^^^ cannot infer type for type parameter `Io`

error[E0282]: type annotations needed
   --> /build/4y8ml4p65982cqjdjjzp3by17lsmgrqn-source/../nix-vendor/actix-http-3.0.0-beta.5/src/client/connector.rs:548:8
    |
548 |     fn call(&self, req: Connect) -> Self::Future {
    |        ^^^^ cannot infer type for type parameter `Io`

error[E0282]: type annotations needed
   --> /build/4y8ml4p65982cqjdjjzp3by17lsmgrqn-source/../nix-vendor/actix-http-3.0.0-beta.5/src/client/connector.rs:546:5
    |
546 |     actix_service::forward_ready!(service);
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type for type parameter `Io`
    |
    = note: this error originates in the macro `actix_service::forward_ready` (in Nightly builds, run with -Z macro-backtrace for more info)

Some errors have detailed explanations: E0282, E0432.
For more information about an error, try `rustc --explain E0282`.
error: could not compile `actix-http` due to 9 previous errors
warning: build failed, waiting for other jobs to finish...
error: build failed
```

--- a/src/builders/rust/build-rust-package/default.nix
+++ b/src/builders/rust/build-rust-package/default.nix
@@ -40,7 +40,7 @@
 
       preConfigure = ''
         mkdir -p $CARGO_HOME
-        mv /build/.cargo/config $CARGO_HOME/config.toml
+        mv ../.cargo/config $CARGO_HOME/config.toml
         ${vendoring.writeGitVendorEntries "vendored-sources"}
         ${vendoring.replaceRelativePathsWithAbsolute}
       '';

--- a/src/builders/rust/build-rust-package/default.nix
+++ b/src/builders/rust/build-rust-package/default.nix
@@ -2,7 +2,7 @@
   lib,
   pkgs,
   ...
-}: {
+} @ topArgs: {
   subsystemAttrs,
   defaultPackageName,
   defaultPackageVersion,
@@ -16,12 +16,15 @@
 } @ args: let
   l = lib // builtins;
 
-  utils = import ../utils.nix args;
-  vendoring = import ../vendor.nix (args // {inherit lib pkgs utils;});
+  utils = import ../utils.nix (args // topArgs);
+  vendoring = import ../vendor.nix (args // topArgs);
 
   buildPackage = pname: version: let
     src = utils.getRootSource pname version;
-    vendorDir = vendoring.vendorDependencies pname version;
+    vendorDir = vendoring.vendoredDependencies;
+    replacePaths =
+      utils.replaceRelativePathsWithAbsolute subsystemAttrs.relPathReplacements;
+    writeGitVendorEntries = vendoring.writeGitVendorEntries "vendored-sources";
 
     cargoBuildFlags = "--package ${pname}";
   in
@@ -41,8 +44,8 @@
       preConfigure = ''
         mkdir -p $CARGO_HOME
         mv ../.cargo/config $CARGO_HOME/config.toml
-        ${vendoring.writeGitVendorEntries "vendored-sources"}
-        ${vendoring.replaceRelativePathsWithAbsolute}
+        ${writeGitVendorEntries}
+        ${replacePaths}
       '';
     });
 in rec {

--- a/src/builders/rust/build-rust-package/default.nix
+++ b/src/builders/rust/build-rust-package/default.nix
@@ -32,7 +32,7 @@
       inherit pname version src;
 
       cargoBuildFlags = cargoBuildFlags;
-      cargoCheckFlags = cargoBuildFlags;
+      cargoTestFlags = cargoBuildFlags;
 
       cargoVendorDir = "../nix-vendor";
 

--- a/src/builders/rust/crane/default.nix
+++ b/src/builders/rust/crane/default.nix
@@ -3,7 +3,7 @@
   pkgs,
   externals,
   ...
-}: {
+} @ topArgs: {
   subsystemAttrs,
   defaultPackageName,
   defaultPackageVersion,
@@ -17,20 +17,24 @@
 } @ args: let
   l = lib // builtins;
 
-  utils = import ../utils.nix args;
-  vendoring = import ../vendor.nix (args // {inherit lib pkgs utils;});
+  utils = import ../utils.nix (args // topArgs);
+  vendoring = import ../vendor.nix (args // topArgs);
 
   crane = externals.crane;
 
   buildPackage = pname: version: let
     src = utils.getRootSource pname version;
-    cargoVendorDir = vendoring.vendorDependencies pname version;
+    cargoVendorDir = vendoring.vendoredDependencies;
+    replacePaths =
+      utils.replaceRelativePathsWithAbsolute subsystemAttrs.relPathReplacements;
+    writeGitVendorEntries = vendoring.writeGitVendorEntries "nix-sources";
+
     postUnpack = ''
       export CARGO_HOME=$(pwd)/.cargo_home
     '';
     preConfigure = ''
-      ${vendoring.writeGitVendorEntries "nix-sources"}
-      ${vendoring.replaceRelativePathsWithAbsolute}
+      ${writeGitVendorEntries}
+      ${replacePaths}
     '';
     # The deps-only derivation will use this as a prefix to the `pname`
     depsNameSuffix = "-deps";

--- a/src/builders/rust/utils.nix
+++ b/src/builders/rust/utils.nix
@@ -2,10 +2,30 @@
   getSourceSpec,
   getSource,
   getRoot,
+  lib,
   ...
-}: rec {
+}: let
+  l = lib // builtins;
+in rec {
+  # Gets the root source for a package
   getRootSource = pname: version: let
     root = getRoot pname version;
   in
     getSource root.pname root.version;
+
+  # Generates a script that replaces relative path dependency paths with absolute
+  # ones, if the path dependency isn't in the source dream2nix provides
+  replaceRelativePathsWithAbsolute = paths: let
+    replacements =
+      l.concatStringsSep
+      " \\\n"
+      (
+        l.mapAttrsToList
+        (rel: abs: "--replace '\"${rel}\"' '\"${abs}\"'")
+        paths
+      );
+  in ''
+    substituteInPlace ./Cargo.toml \
+      ${replacements}
+  '';
 }

--- a/src/builders/rust/vendor.nix
+++ b/src/builders/rust/vendor.nix
@@ -1,43 +1,24 @@
 {
   lib,
   pkgs,
-  getRoot,
   getSource,
   getSourceSpec,
-  getDependencies,
-  getCyclicDependencies,
   subsystemAttrs,
+  dreamLock,
   ...
 }: let
   l = lib // builtins;
 
-  isCyclic = cyclic: dep:
-    l.any (odep: dep.name == odep.name && dep.version == odep.version) cyclic;
-
-  getUncyclicDependencies = cyclic: deps:
-    l.map
-    (dep: getAllTransitiveDependencies dep.name dep.version)
-    (l.filter (dep: !(isCyclic cyclic dep)) deps);
-
-  getAllTransitiveUncyclicDependencies = pname: version: let
-    cyclic = getCyclicDependencies pname version;
-    direct = getDependencies pname version;
-  in
-    l.unique (l.flatten (direct ++ (getUncyclicDependencies cyclic direct)));
-
-  getAllTransitiveDependencies = pname: version: let
-    direct = getDependencies pname version;
-    cyclic = getCyclicDependencies pname version;
-  in
-    l.unique (l.flatten (
-      direct
-      ++ (getUncyclicDependencies cyclic direct)
-      ++ (
-        l.map
-        (dep: getAllTransitiveUncyclicDependencies dep.name dep.version)
-        (l.filter (isCyclic cyclic) direct)
+  allDependencies =
+    l.flatten
+    (
+      l.mapAttrsToList
+      (
+        name: versions:
+          l.map (version: {inherit name version;}) (l.attrNames versions)
       )
-    ));
+      dreamLock.dependencies
+    );
 in rec {
   # Generates a shell script that writes git vendor entries to .cargo/config.
   # `replaceWith` is the name of the vendored source(s) to use.
@@ -56,11 +37,8 @@ in rec {
     EOF
   '';
 
-  # Vendor a package's dependencies like how `cargo vendor` would do,
-  # so we can use it with `cargo`.
-  vendorPackageDependencies = pname: version: let
-    deps = getAllTransitiveDependencies pname version;
-
+  # Vendors the dependencies passed as Cargo expects them
+  vendorDependencies = deps: let
     makeSource = dep: let
       path = getSource dep.name dep.version;
       spec = getSourceSpec dep.name dep.version;
@@ -107,7 +85,7 @@ in rec {
         ${l.optionalString isGit "printf '{\"files\":{},\"package\":null}' > \"$out/${source.name}/.cargo-checksum.json\""}
       '';
   in
-    pkgs.runCommand "vendor-${pname}-${version}" {} ''
+    pkgs.runCommand "vendor" {} ''
       mkdir -p $out
 
       ${
@@ -117,26 +95,6 @@ in rec {
       }
     '';
 
-  # Vendors a package's roots dependencies.
-  vendorDependencies = pname: version: let
-    root = getRoot pname version;
-  in
-    vendorPackageDependencies root.pname root.version;
-
-  # Generates a script that replaces relative path dependency paths
-  # with absolute ones, if the path dependency isn't in the source
-  # dream2nix provides
-  replaceRelativePathsWithAbsolute = let
-    replacements =
-      l.concatStringsSep
-      " \\\n"
-      (
-        l.mapAttrsToList
-        (rel: abs: "--replace '\"${rel}\"' '\"${abs}\"'")
-        subsystemAttrs.replacePathsWithAbsolute
-      );
-  in ''
-    substituteInPlace ./Cargo.toml \
-      ${replacements}
-  '';
+  # All dependencies in the Cargo.lock file, vendored
+  vendoredDependencies = vendorDependencies allDependencies;
 }

--- a/src/default.nix
+++ b/src/default.nix
@@ -289,6 +289,7 @@
         inherit
           buildPackageWithOtherBuilder
           produceDerivation
+          dreamLock
           ;
 
         inherit

--- a/src/translators/rust/pure/cargo-lock/default.nix
+++ b/src/translators/rust/pure/cargo-lock/default.nix
@@ -158,7 +158,7 @@ in {
       # The structure of this should be defined in:
       #   ./src/specifications/{subsystem}
       subsystemAttrs = rec {
-        replacePathsWithAbsolute = let
+        relPathReplacements = let
           # Extract dependencies from the Cargo.toml of the
           # package we are currently building
           tomlDeps =
@@ -231,13 +231,12 @@ in {
           sourceType = getSourceTypeFrom rawObj;
           sourceConstructors = {
             path = dependencyObject: let
-              findToml =
+              findCrate =
                 l.findFirst
                 (
                   crate:
-                    crate.name
-                    == dependencyObject.name
-                    && crate.version == dependencyObject.version
+                    (crate.name == dependencyObject.name)
+                    && (crate.version == dependencyObject.version)
                 )
                 null;
               workspaceCrates =
@@ -249,34 +248,33 @@ in {
                   }
                 )
                 cargoPackages;
-              workspaceToml = findToml workspaceCrates;
-              nonWorkspaceToml = findToml (subsystemInfo.crates or []);
+              workspaceCrate = findCrate workspaceCrates;
+              nonWorkspaceCrate = findCrate (subsystemInfo.crates or []);
             in
               if
-                package.name
-                == dependencyObject.name
-                && package.version == dependencyObject.version
+                (package.name == dependencyObject.name)
+                && (package.version == dependencyObject.version)
               then
                 dlib.construct.pathSource {
                   path = projectSource;
                   rootName = null;
                   rootVersion = null;
                 }
-              else if workspaceToml != null
+              else if workspaceCrate != null
               then
                 dlib.construct.pathSource {
-                  path = workspaceToml.relPath;
+                  path = workspaceCrate.relPath;
                   rootName = package.name;
                   rootVersion = package.version;
                 }
-              else if nonWorkspaceToml != null
+              else if nonWorkspaceCrate != null
               then
                 dlib.construct.pathSource {
-                  path = "${rootSource}/${nonWorkspaceToml.relPath}";
+                  path = dlib.sanitizePath "${rootSource}/${nonWorkspaceCrate.relPath}";
                   rootName = null;
                   rootVersion = null;
                 }
-              else throw "could not find crate ${dependencyObject.name}";
+              else throw "could not find crate '${dependencyObject.name}-${dependencyObject.version}'";
 
             git = dependencyObject: let
               parsed = parseGitSource dependencyObject.source;

